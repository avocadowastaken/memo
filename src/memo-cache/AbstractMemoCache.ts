export abstract class AbstractMemoCache<TKey, TValue> {
  public abstract get(key: TKey): TValue | undefined;

  public abstract prime(key: TKey, value: TValue): this;

  public abstract clear(key: TKey): this;

  public abstract clearAll(): this;
}
