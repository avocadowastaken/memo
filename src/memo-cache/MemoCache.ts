import { AbstractMemoCache } from "./AbstractMemoCache";
import { LRUCache } from "./internal/LRUCache";
import { MapCache } from "./internal/MapCache";

export type MemoCacheKeyType<TKey> = TKey | string | number;

export type MemoCacheKeyFn<TKey> = (key: TKey) => MemoCacheKeyType<TKey>;

export interface MemoCacheOptions<TKey> {
  readonly maxSize?: number;
  readonly cacheKeyFn?: MemoCacheKeyFn<TKey>;

  readonly expireAfterWrite?: number;
  readonly expireAfterAccess?: number;
}

interface MemoCacheNode<TValue> {
  value: TValue;
  expiresAt: number;
}

export class MemoCache<TKey, TValue> extends AbstractMemoCache<TKey, TValue> {
  protected readonly cache: AbstractMemoCache<
    MemoCacheKeyType<TKey>,
    MemoCacheNode<TValue>
  >;

  protected readonly cacheKeyFn: MemoCacheKeyFn<TKey>;

  protected readonly expireAfterWrite: number;

  protected readonly expireAfterAccess: number;

  public constructor({
    maxSize,

    expireAfterWrite = 0,
    expireAfterAccess = 0,
    cacheKeyFn = (key: TKey) => key,
  }: MemoCacheOptions<TKey> = {}) {
    super();

    this.cacheKeyFn = cacheKeyFn;
    this.expireAfterWrite = expireAfterWrite;
    this.expireAfterAccess = expireAfterAccess;

    this.cache = !maxSize ? new MapCache() : new LRUCache({ maxSize });
  }

  public get(key: TKey): TValue | undefined {
    const node = this.cache.get(this.cacheKeyFn(key));

    if (node) {
      if (node.expiresAt > Date.now()) {
        if (this.expireAfterAccess) {
          node.expiresAt = Math.max(
            node.expiresAt,
            Date.now() + this.expireAfterAccess,
          );
        }

        return node.value;
      }

      this.clear(key);
    }
  }

  public prime(key: TKey, value: TValue): this {
    const expiresAfter = Math.max(
      this.expireAfterWrite || 0,
      this.expireAfterAccess || 0,
    );

    const node = {
      value,
      expiresAt: expiresAfter === 0 ? Infinity : expiresAfter + Date.now(),
    };

    this.cache.prime(this.cacheKeyFn(key), node);

    return this;
  }

  public clear(key: TKey): this {
    this.cache.clear(this.cacheKeyFn(key));

    return this;
  }

  public clearAll(): this {
    this.cache.clearAll();

    return this;
  }
}
