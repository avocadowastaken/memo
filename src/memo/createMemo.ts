import { assertFn } from "../internal/assertFn";
import { MemoCache, MemoCacheOptions } from "../memo-cache/MemoCache";

export interface MemoFn<TKey, TValue> {
  (key: TKey): TValue;

  readonly cache: MemoCache<TKey, TValue>;
}

export function createMemo<TKey, TValue>(
  fn: (key: TKey) => TValue,
  options?: MemoCacheOptions<TKey>,
): MemoFn<TKey, TValue> {
  assertFn(fn);

  const cache = new MemoCache<TKey, TValue>(options);

  memo.cache = cache;

  return memo;
  function memo(key: TKey): TValue {
    let value = cache.get(key);

    if (value === undefined) {
      value = fn(key);
      cache.prime(key, value);
    }

    return value;
  }
}
