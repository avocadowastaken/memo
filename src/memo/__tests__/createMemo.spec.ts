import { MemoCache } from "../../memo-cache/MemoCache";
import { createMemo } from "../createMemo";

function createCounter<K>() {
  const cache = new Map<K, number>();

  return (key: K): number => {
    const count = (cache.get(key) || 0) + 1;

    cache.set(key, count);

    return count;
  };
}

it("validates 'fn'", () => {
  expect(() => createMemo(null as any)).toThrowErrorMatchingSnapshot();
});

it("exposes cache", () => {
  const memo = createMemo(createCounter());

  expect(memo.cache).toBeInstanceOf(MemoCache);
});

it("evaluates 'fn'", () => {
  const fn = jest.fn(createCounter());
  const memo = createMemo(fn);

  expect(fn).toHaveBeenCalledTimes(0);

  expect(memo(1)).toBe(1);
  expect(fn).toHaveBeenCalledTimes(1);

  expect(memo(1)).toBe(1);
  expect(fn).toHaveBeenCalledTimes(1);
});
