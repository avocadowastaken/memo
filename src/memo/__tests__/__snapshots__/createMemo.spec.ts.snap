// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validates 'fn' 1`] = `"Memo: 'fn' expected to be a 'function'."`;
